### PR TITLE
[FEATURE] Device code authentication flow through CLI

### PR DESCRIPTION
**Pull request by celian-garcia**
# Description

Implementation of the device auth flow according to the [specs](https://github.com/perses/perses/blob/main/docs/design-docs/authentication.md#-login-from-external-oidc-or-oauth20-provider-through-percli-command-line) .
After this PR we can sign in with an external provider through percli

# Screenshots

image
image
image

# Checklist

- [x] Pull request has a descriptive title and context useful to a reviewer.
- [x] Pull request title follows the `[<catalog_entry>] <commit message>` naming convention using one of the
  following `catalog_entry` values: `FEATURE`, `ENHANCEMENT`, `BUGFIX`, `BREAKINGCHANGE`, `DOC`,`IGNORE`.
- [x] All commits have [DCO signoffs](https://github.com/probot/dco#how-it-works).

### CODE REVIEW

File: pkg/model/api/config/authentication.go [hunk 1]
```diff
@@ -28,6 +28,12 @@ const (
 	DefaultRefreshTokenTTL = time.Hour * 24
 )
 
+type DeviceCode struct {
+	Disabled     bool          `json:"disabled" yaml:"disabled"`
```
Comment (celian-garcia) [inline pkg/model/api/config/authentication.go:32]: Remove the Disabled

File: internal/api/impl/auth/oidc.go [hunk 2]
```diff
@@ -27,14 +28,36 @@ import (
...
 	v1 "github.com/perses/perses/pkg/model/api/v1"
 	"github.com/sirupsen/logrus"
+	"github.com/zitadel/oidc/v3/pkg/client"
 	"github.com/zitadel/oidc/v3/pkg/client/rp"
 	httphelper "github.com/zitadel/oidc/v3/pkg/http"
 	"github.com/zitadel/oidc/v3/pkg/oidc"
+	"golang.org/x/oauth2"
 )
 
+// grantType is a subset of the grant types of OIDC library.
```
Comment (celian-garcia) [inline internal/api/impl/auth/oidc.go:42]: Move it to a common place, since it's used by oidc and oauth files

File: internal/api/impl/auth/oidc.go [hunk 3]
```diff
@@ -92,12 +124,11 @@ func newOIDCEndpoint(provider config.OIDCProvider, jwt crypto.JWT, dao user.DAO)
 	//   from cookie to use them before deleting the cookies.
 	key := securecookie.GenerateRandomKey(16)
 	cookieHandler := httphelper.NewCookieHandler(key, key)
-	client := &http.Client{
```
Comment (celian-garcia) [inline internal/api/impl/auth/oidc.go:95]: polluting unnecessary change

File: internal/api/impl/auth/oauth.go [hunk 4]
```diff
@@ -282,71 +300,198 @@ func (e *oAuthEndpoint) codeExchangeHandler(ctx echo.Context) error {
...
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to exchange client credentials for token")
+		}
+	default:
+		return echo.NewHTTPError(http.StatusBadRequest, "Unsupported grant type")
+	}
 
-	username := entity.GetMetadata().GetName()
-	_, err = e.tokenManagement.accessToken(username, ctx.SetCookie)
+	uInfo, err := e.requestUserInfo(ctx, accessToken)
+	if err != nil {
```
Comment (celian-garcia) [inline internal/api/impl/auth/oauth.go:370]: We'd probably need to create a more fine-grained error management to differentiate the errors that needs to be considered as expected (to be tried again) or fatal (a next shot will not be better)

File: internal/api/impl/auth/oauth.go [hunk 5]
```diff
@@ -282,71 +301,205 @@ func (e *oAuthEndpoint) codeExchangeHandler(ctx echo.Context) error {
 		return apiinterface.InternalError
 	}
 
-	uInfoBody, err := e.requestUserInfo(ctx, token)
+	uInfo, err := e.requestUserInfo(ctx, token)
 	if err != nil {
 		e.logWithError(err).Warn("An error occurred while requesting user infos. User infos will be ignored.")
```
Comment (celian-garcia) [inline internal/api/impl/auth/oauth.go:306]: Change Warn to Error since the user infos is now mandatory

File: internal/api/impl/auth/oauth.go (hunk 4 above)
Comment (celian-garcia) [reply internal/api/impl/auth/oauth.go:374]: I materialized it into a comment. 
```go
                        // Note that for sake of simplicity, it doesn't respect the Oauth 2.0 RFC: (https://www.rfc-editor.org/rfc/rfc8628#section-3.5)
			// For example:
			// - we should differentiate the errors that means "try again later" from the others. (error: "authorization_pending")
			// - we should propagate the message possibly saying to slow down the request (error: "slow_down")
```

Not sure if we do the effort of comply fully with OAuth 2.0 specs on that. If yes maybe a larger work would be necessary to comply in all the auth endpoints error response... which is quite a lot of work oO

File: internal/cli/cmd/login/login.go [hunk 6]
```diff
@@ -125,10 +126,31 @@ func (o *option) authAndSetToken() error {
...
+	// Display the user code and verification URL
+	fmt.Printf("Go to %s and enter this user code: %s\n", deviceCodeResponse.VerificationURI, deviceCodeResponse.UserCode)
+
+	// Poll for an access token
+	for {
+		time.Sleep(time.Duration(deviceCodeResponse.Interval) * time.Second)
+
+		tokenResponse, err := o.apiClient.Auth().PollDeviceCodeFlow(string(o.externalAuthKind), o.externalAuthProvider, deviceCodeResponse.DeviceCode)
+		if err != nil {
+			fmt.Println("Waiting for user to authorize the application...")
```
[the remark this answers is not in the document]
Comment (celian-garcia) [reply internal/cli/cmd/login/login.go]: Is it possible to use HandleString without newline at the end ? I changed a bit that code to not write a new line every time but just add a doc. It is painful that the lines move when we try to copy paste the code..

File: internal/api/impl/auth/oidc.go [hunk 7]
```diff
@@ -114,22 +127,44 @@ func newOIDCEndpoint(provider config.OIDCProvider, jwt crypto.JWT, dao user.DAO)
...
+
+	// Add routes for device code flow and token exchange
+	g.POST(fmt.Sprintf("/%s/%s/%s", utils.AuthKindOIDC, e.slugID, utils.PathDeviceCode), e.deviceCodeHandler, true)
+	g.POST(fmt.Sprintf("/%s/%s/%s", utils.AuthKindOIDC, e.slugID, utils.PathToken), e.tokenHandler, true)
 }
 
+// buildAuthHandler builds the http handler on Perses side that triggers the "Authorization Code"
+// flow to the oauth 2.0 provider.
+// It will redirect the user to the provider's "auth" url.
 func (e *oIDCEndpoint) buildAuthHandler() echo.HandlerFunc {
```
[the remark this answers is not in the document]
Comment (celian-garcia) [reply internal/api/impl/auth/oidc.go:168]: not sure, I think that historically there was more code outside the returned func. But I see that now the main part is insde so I can transform

File: internal/api/impl/auth/oauth.go [hunk 8]
```diff
@@ -282,71 +301,213 @@ func (e *oAuthEndpoint) codeExchangeHandler(ctx echo.Context) error {
...
+	case GrantTypeDeviceCode:
+		accessToken, err = e.retrieveDeviceAccessToken(ctx.Request().Context(), reqBody.DeviceCode)
+		if err != nil {
+			// (We log a warning as the failure means most of the time that the user didn't authorize the app yet)
+			e.logWithError(err).Warn("Failed to exchange device code for token")
+			// Note that for sake of simplicity, it doesn't respect the Oauth 2.0 RFC: (https://www.rfc-editor.org/rfc/rfc8628#section-3.5)
+			// For example:
+			// - we should differentiate the errors that means "try again later" from the others. (error: "authorization_pending")
+			// - we should propagate the message possibly saying to slow down the request (error: "slow_down")
+			return echo.NewHTTPError(http.StatusInternalServerError, "Failed to exchange device code for token")
```
[the remark this answers is not in the document]
Comment (celian-garcia) [reply internal/api/impl/auth/oauth.go:365]: Checking the oauth 2.0 specs, it seems that we need to return a 400 with an ``error`` field in the JSON containing a value from a limited list. 
The thing is there is a lot of possible value in that list, and if we do it properly, we should respect as much as possible the values returned by the provider, and some providers are not respecting it.

I confess I took a shortcut here but we need to speak to know what do we do exactly.

File: internal/cli/cmd/login/login.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (Nexucis) [reply internal/cli/cmd/login/login.go]: I don't get what you would like :(. 

HandleString is doing the same thing than `fmt.Println` it's just using the writer instead of a default one

[the remark this answers is not in the document]
Comment (celian-garcia) [reply internal/cli/cmd/login/login.go]: I changed the code since. I want to do fmt.Print now :P

[the remark this answers is not in the document]
Comment (celian-garcia) [reply internal/cli/cmd/login/login.go]: Nevermind I will write the line only once and that will be readable enough.

File: internal/api/impl/auth/oauth.go (hunk 4 above)
Comment (celian-garcia) [reply internal/api/impl/auth/oauth.go:378]: As discussed, I propose to move the target of this PR to a feature branch ``feature/extend-oauth`` and I will make a separate PR to make fine-grain ``/token`` error management. Following the OAuth 2.0 RFC. 
I will also finalize the client credentials flow in that branch, through another PR.

File: internal/api/impl/auth/oauth.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (celian-garcia) [reply internal/api/impl/auth/oauth.go:365]: As discussed, I propose to move the target of this PR to a feature branch feature/extend-oauth and I will make a separate PR to make fine-grain /token error management. Following the OAuth 2.0 RFC.
I will also finalize the client credentials flow in that branch, through another PR.

**Review by Nexucis — APPROVED** [review]
